### PR TITLE
refactor: remove usage of swc's `SourceMap` and `SourceFile`

### LINKED ISSUE #790: Remove usage of `SourceMap` and `SourceFile`
I'm going to work on https://github.com/dprint/dprint-swc-ecma-ast-view/issues/28 first, but we should remove the usage of these as there is no need to have them and they do not implement `Sync`.

This is part of the work for reducing extra parses in Deno's language server.

### PR DESCRIPTION
**Pull request by dsherret**
Closes #790

FYI @magurotuna

### CODE REVIEW

File: src/ast_parser.rs [hunk 1]
```diff
@@ -133,83 +102,60 @@ pub(crate) struct AstParser {
...
+      SwcDiagnostic::from_swc_error(file_name, &source_file, err)
+    })?;
 
-    let parse_result = parser.parse_program().map_err(move |err| {
-      let mut diagnostic_builder = err.into_diagnostic(&self.handler);
-      diagnostic_builder.emit();
-      SwcDiagnosticBuffer::from_swc_error(buffered_err, self)
+    let program = swc_common::GLOBALS.set(&self.globals, || {
+      program.fold_with(&mut ts_resolver(self.top_level_mark))
     });
```
Comment (dsherret) [inline src/ast_parser.rs:144]: What does this do?

File: src/diagnostic.rs [hunk 2]
```diff
@@ -1,25 +1,23 @@
...
 use serde::Serialize;
-use std::convert::TryInto;
 
 #[derive(Debug, Clone, Copy, PartialEq, Serialize)]
 #[serde(rename_all = "camelCase")]
 pub struct Position {
-  pub line: usize,
-  pub col: usize,
+  pub line_index: usize,
+  pub column_index: usize,
```
Comment (dsherret) [inline src/diagnostic.rs:8]: This should all be 0-index now except when displayed or in the tests. (Still that weird 1-indexed line and 0-indexed column in the tests... maybe we could change that later to both 1-indexed over there)

File: src/rules.rs [hunk 3]
```diff
@@ -55,7 +55,6 @@ pub mod no_invalid_regexp;
 pub mod no_invalid_triple_slash_reference;
 pub mod no_irregular_whitespace;
 pub mod no_misused_new;
-pub mod no_mixed_spaces_and_tabs;
```
Comment (dsherret) [inline src/rules.rs:58]: Should this rule exist? I think it's a formatting concern of `deno fmt` and not `deno lint` so I removed it, but I mainly removed it in this PR because it would be a bit of work to update it to work now (because it was using `source_map().span_to_lines(...)`).

File: examples/dlint/js.rs [hunk 4]
```diff
@@ -149,7 +149,7 @@ impl Plugin for JsRuleRunner {
   fn run(
     &mut self,
     context: &mut Context,
-    program: Program,
+    program: ProgramRef,
```
Comment (dsherret) [inline examples/dlint/js.rs:152]: The plugins now take a `ProgramRef` because that's what might be provided to `parse_with_ast`. This also eliminates cloning of a whole AST potentially a bunch of times.

File: src/diagnostic.rs (hunk 2 above)
Comment (bartlomieju) [reply src/diagnostic.rs:8]: Keep in mind that this is used by LSP/vscode_deno, so changing it might break the former

**Review by bartlomieju — COMMENTED** [review]
This looks good to me, I'll let @magurotuna review it as well

File: examples/dlint/js.rs [hunk 5]
```diff
@@ -11,12 +11,12 @@ use deno_core::ZeroCopyBuf;
 use deno_lint::context::Context;
 use deno_lint::control_flow::ControlFlow;
 use deno_lint::linter::Plugin;
+use deno_lint::rules::ProgramRef;
```
[the remark this answers is not in the document]
Comment (dsherret) [reply examples/dlint/js.rs:14]: Actually, I think it would be ok to remove the `ProgramRef` now and use the one in `dprint_swc_ecma_ast_view`.

**Comment by magurotuna** [comment]
I will review at night (in about 10 hours).

File: src/diagnostic.rs (hunk 2 above)
Comment (magurotuna) [reply src/diagnostic.rs:8]: I think we shouldn't use `usize` everywhere to represent line/column numbers because overuse of `usize` may result in developers' confusion in whether this type expresses 0-indexed value that is internally used or 1-indexed value that is for displaying diagnostics to end users. Instead, we could create new types, one representing internal values and the other values to be displayed.

---

Oh I found https://github.com/dsherret/text_lines which provides new types I described above. Maybe we could utilize it later on.

**Review by magurotuna — COMMENTED** [review]
Looks good overall, but one point I want to ask is whether we should replace our `ProgramRef` with the one in `dprint_swc_ecma_ast_view`.

File: src/ast_parser.rs (hunk 1 above)
Comment (bartlomieju) [reply src/ast_parser.rs:144]: This is used to apply proper "syntax context" to all AST elements - when SWC performs transforms/folding it might change some of those context and "ts_resolver" ensures that all elements end up in proper lexical scope.

File: src/diagnostic.rs (hunk 2 above)
Comment (dsherret) [reply src/diagnostic.rs:8]: I'll add some comments to indicate it's 0-indexed. I think using `usize` is useful here because then you can index these types into an array of lines or a characters.

Yup, we could use text_lines in the future.

File: Cargo.toml [hunk 6]
```diff
@@ -21,6 +21,7 @@ default = []
 docs = []
 
 [dependencies]
+ast_view = { version = "0.29.3", package = "dprint-swc-ecma-ast-view" }
```
Comment (dsherret) [inline Cargo.toml:24]: The name `dprint_swc_ecma_ast_view` is way too long. Is this ok to do instead?

File: examples/dlint/js.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsherret) [reply examples/dlint/js.rs]: Done now.

File: Cargo.toml (hunk 6 above)
Comment (magurotuna) [reply Cargo.toml:24]: Fantastic

File: src/diagnostic.rs (hunk 2 above)
Comment (magurotuna) [reply src/diagnostic.rs:10]: I see, that makes sense!

File: src/rules.rs [hunk 7]
```diff
@@ -1,6 +1,8 @@
 // Copyright 2020-2021 the Deno authors. All rights reserved. MIT license.
 use crate::context::Context;
-use dprint_swc_ecma_ast_view::Program as ProgramView;
+
+pub use ast_view::Program;
+pub use ast_view::ProgramRef;
```
Comment (magurotuna) [inline src/rules.rs:5]: How about moving these `pub use` to the crate root (i.e. src/lib.rs)?
These two types are now included in the public interface of this crate, and also it'd be more intuitive for them to live and re-exported in the crate root than to be here as they don't seem to be related to "rules" very much.

**Review by bartlomieju — APPROVED** [review]

File: src/rules.rs (hunk 7 above)
Comment (dsherret) [reply src/rules.rs]: Done in c9e2996e0c362130deb79cfc994d9041dde560bb.

**Review by magurotuna — APPROVED** [review]
LGTM 👍
